xarray/core/utils.py

```diff
@@ -290,172 +319,130 @@ def ordered_dict_intersection(first_dict, second_dict, compat=equivalent):
 
 
-class SortedKeysDict(MutableMapping, SingleSlotPickleMixin):
+class SortedKeysDict(MutableMapping[K, V], SingleSlotPickleMixin):
     """An wrapper for dictionary-like objects that always iterates over its
     items in sorted order by key but is otherwise equivalent to the underlying
     mapping.
     """
     __slots__ = ['mapping']
 
-    def __init__(self, mapping=None):
+    def __init__(self, mapping: Optional[MutableMapping[K, V]] = None) -> None:
         self.mapping = {} if mapping is None else mapping
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: K) -> V:
         return self.mapping[key]
 
-    def __setitem__(self, key, value):
+    def __setitem__(self, key: K, value: V) -> None:
         self.mapping[key] = value
 
-    def __delitem__(self, key):
+    def __delitem__(self, key: K) -> None:
         del self.mapping[key]
 
-    def __iter__(self):
+    def __iter__(self) -> Iterator[K]:
         return iter(sorted(self.mapping))
 
-    def __len__(self):
+    def __len__(self) -> int:
         return len(self.mapping)
 
-    def __contains__(self, key):
+    def __contains__(self, key: object) -> bool:
         return key in self.mapping
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return '%s(%r)' % (type(self).__name__, self.mapping)
 
-    def copy(self):
-        return type(self)(self.mapping.copy())
-
-
-class ChainMap(MutableMapping, SingleSlotPickleMixin):
-    """Partial backport of collections.ChainMap from Python>=3.3
-
-    Don't return this from any public APIs, since some of the public methods
-    for a MutableMapping are missing (they will raise a NotImplementedError)
-    """
-    __slots__ = ['maps']
-
-    def __init__(self, *maps):
-        self.maps = maps
-
-    def __getitem__(self, key):
-        for mapping in self.maps:
-            try:
-                return mapping[key]
-            except KeyError:
-                pass
-        raise KeyError(key)
 
-    def __setitem__(self, key, value):
-        self.maps[0][key] = value
-
-    def __delitem__(self, value):  # pragma: no cover
-        raise NotImplementedError
-
-    def __iter__(self):
-        seen = set()
-        for mapping in self.maps:
-            for item in mapping:
-                if item not in seen:
-                    yield item
-                    seen.add(item)
-
-    def __len__(self):
-        raise len(iter(self))
-
-
-class OrderedSet(MutableSet):
+class OrderedSet(MutableSet[T]):
     """A simple ordered set.
 
     The API matches the builtin set, but it preserves insertion order of
     elements, like an OrderedDict.
     """
-
-    def __init__(self, values=None):
-        self._ordered_dict = OrderedDict()
+    def __init__(self, values: Optional[AbstractSet[T]] = None) -> None:
+        self._ordered_dict = OrderedDict()  # type: MutableMapping[T, None]
         if values is not None:
-            self |= values
+            self.__ior__(values)
```

Comment:
Does mypy not recognize `|=` as valid? That's usually preferred since that's the operator `__ior__` corresponds to.

If not, that's OK but please leave a comment.

Reply:
Neither mypy nor PyCharm are happy with it.
It has to do with the signature of ``__ior__``:
```
def __ior__(self: MutableSet[T1], other: AbstractSet[T2]) -> MutableSet[Union[T1, T2]]:
    ...
```
When you do ``self |= other``, both PyCharm and mypy understand that you're changing the type of self in place, in other word they treat it to be the same as ``self = self | other``.
Calling ``__ior__`` directly bypasses that. 
This seems to be like a limitation of both tools like in our specific cases we explicitly specified that other is an ``AbstractSet[T1]``. I'll revert to |= and just disable the type checking.

Reply:
OK, I'm happy with a comment, too. It's also not a big deal to disable type checking on particular lines when we're sure the overall method/function signature is right.

Reply:
I'm not entirely sure, but It might help to annotate `self` here